<a href="https://colab.research.google.com/github/JuanXo375/AlgorithmsUN2023I/blob/main/Lab4/jcarrenoarGroup1Lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [6]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [7]:
!g++ APlusB.cpp -o APlusB
!./APlusB

4 8
12

### Test input output files

In [8]:
%%writefile input1.txt
4 8

Overwriting input1.txt


In [9]:
%%writefile output1.txt
12

Writing output1.txt


In [10]:
%%writefile input2.txt
-5 9

Overwriting input2.txt


In [11]:
%%writefile output2.txt
4

Writing output2.txt


In [12]:
!./APlusB < input1.txt

12

In [13]:
!./APlusB < input2.txt

4

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [14]:
!grep -xvFf output1.txt output1.txt 

In [15]:
!grep -xvFf output1.txt output2.txt 

4


In [16]:
!grep -xvFf output2.txt output1.txt 

12


In [17]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [18]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [19]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [20]:
!python sum.py

4 8
12


In [21]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [22]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [23]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product.cpp


In [24]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!./max_pairwise_product

6
12 538 2135 6845 2654 125
18166630


### Test 1: Small numbers

In [25]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [26]:
!./max_pairwise_product < input1.txt

90


### Test 2: largest input numbers 200000 200000 in the input

In [27]:
%%writefile input2.txt
6
1 3 200000 9 200000 23

Overwriting input2.txt


In [28]:
!./max_pairwise_product < input2.txt

1345294336


Changing the variables to long long

In [29]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long.cpp


In [30]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!./max_pairwise_product_long < input1.txt
!./max_pairwise_product_long < input2.txt

90
40000000000


### Test 3  stress tests 20  times  random n in [2, 20] 

In [31]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>


using namespace std;

long long MaxPairwiseProduct(const vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
			       aux_product =  (long long) numbers[first] * numbers[second];
            if ( max_product < aux_product)  {
				        max_product = aux_product;
             }
        }
    }

    return max_product;
}

int main() {
    srand(time(NULL)); // inicializar semilla aleatoria
    for (int i = 0; i < 20; i++) {
        int n = rand() % 19 + 2; // tamaño aleatorio del arreglo entre [2, 20]
        vector<int> numbers(n);
        for (int j = 0; j < n; j++) {
            numbers[j] = rand() % 1000; // números aleatorios entre [0, 999]
        }
        cout << MaxPairwiseProduct(numbers) << endl;
    }
    return 0;
}



Writing test3.cpp


In [32]:
!g++ test3.cpp -o test3
!./test3

915488
682945
608076
726375
901394
19364
747968
744861
643386
398160
905332
463420
654110
665483
680067
638308
529818
525264
744576
819896


### Test 4  time test  measure the execution time running with the the largest worst case file (200000 increasing numbers, 200000 and 200000 the end) 

In [33]:
%%writefile Test2Write.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input3.txt");

 MyFile << 200000 <<"\n";  // n 
 for (int i=1; i < 200001; i++){
     MyFile << i <<" ";
 }
  
  // Close the file
  MyFile.close();
}

Overwriting Test2Write.cpp


In [34]:
!g++ Test2Write.cpp -o Test2Write
!./Test2Write

Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [35]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input3.txt

39999800000

real	2m22.365s
user	2m19.783s
sys	0m0.069s


##### Calculate  the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line 

```cpp
 int main() {                         //N inst exec by line executed  
   
    int n;                            //1                      
    std::cin >> n;                    //1
                                              
    std::vector<int> numbers(n);      //n    init vector
    .
    .
 }

```

In [36]:
%%writefile test_init_vectors.cpp
#include <iostream>
#include <vector>
#include <algorithm>
int main() {                                                               
    int n = 10;                         //1
    std::vector<int> numbers(n);        //n                                    
    for (int i = 0; i < n; ++i) {       //1+2n                                    
        std::cout << numbers[i];        //n                                     
    }

    std::cout << "\n";                  //1     
    return 0;                                                              
}                                       //1+n+1+2n+n+1

Writing test_init_vectors.cpp


In [37]:
!g++ test_init_vectors.cpp -o test_init_vectors
!./test_init_vectors

0000000000


In [38]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed 
    long long max_product = 0LL;                                             //1                                          
	  long long aux_product = 0LL;                                             //1 
    int n = numbers.size();                                                  //1 
                                                                             
    for (int first = 0; first < n; ++first) {                                //1+2n  
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n 
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            if (max_product < aux_product)  {                                //(n-1)n/2 
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
             }
        }
    }
                                                                             // -------------- 
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                              //N inst exec by line executed 
    int n;                                                //1  
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // -------------- 
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]


Overwriting max_pairwise_product_long.cpp


Count executed intructions in code and verify with inputs 1, 2 and 3

In [39]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;                                                       //1
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                                //1+2n  
        execints += 2;

        long long m = 0LL;

        execints++;                                    
        for (int second = first + 1; second < n; ++second) {                 //n+ (n-1)n 
            execints += 2;                                                

			      aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2 
            execints++; 

            execints++; 
            if ( max_product < aux_product)  {                               //(n-1)n/2 
                       
				        max_product = aux_product;                                   //[0, (n-1)n/2] 
                execints++; 
                m++;
             }
        }
        if ( (first <= 10) || (n- first <= 10)) std::cout << m << "\n";
    }
                                                                            // -------------- 
    return max_product;                                                     // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }
    
    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";                       
    return 0;                                                // -------------- 
}                                                            // 9 + 7n + [2(n-1)n, 2.5 (n-1)n]

Overwriting max_pairwise_product_long_execints.cpp


In [40]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!./max_pairwise_product_long_execints < input1.txt
!./max_pairwise_product_long_execints < input2.txt
!./max_pairwise_product_long_execints < input3.txt

9
5
4
3
2
2
2
2
1
0
289
2
1
2
0
0
0
116
199999
100000
66667
50000
40000
33334
28572
25000
22223
20000
18182
2
2
2
2
2
2
2
2
1
0
80003672077


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T = 9 + 7n + [2(n-1)n, 2.5 (n-1)n]$$

It depends on the size of the lists and  the order of the list.

For the input1.txt 
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 7(10) + 2(9)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 289 $$

For the input2.txt 
```
6
1 3 200000 9 200000 23
```
$$T = 9 + 7(6) + 2(5)(6)  +  2 + 1 + 2 + 0 + 0 + 0 = 116 $$


For the input3.txt that is the worst case (n=200000 and list in incresing order)

```
200000
1 2 3 4 5 6 7 8 9 10 ... 199998 199999 200000
```

$$
\begin{array}{ll}
T &= 9 + 7(200000) + (2)(199999)(200000)  + (199999 +
100000 + 66667 +50000 \cdots 2 + 2 + 1)\\
&= 80003672077  \\
  &< 9 + 7(200000) + (2.5)(199999)(200000) \\
  & = 100000900009
\end{array}
  $$

Calculate estimated time by instruction and number of instructions by second

In [41]:
# Estimated time by instruction

128/80003672077 


1.5999265618308827e-09

In [42]:
# Estimated number of instructions by second

1/(128/80003672077) 

625028688.1015625

### Efficient solution

Add in each line a coment with the number of intstrcution that are executed in that line

In [121]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<long long>& numbers, long long &execints) {              //N inst exec by line executed 
    
    long long m = 0LL;
    
    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                               //1 
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                             //1+2n  
        execints += 2;

        int hook = max_product;                                           //n
        execints++;

        max_product = std::max(max_product,numbers[first]);               //n
        execints++;

        execints++;
        execints++;
        if(hook != max_product || max_product == numbers[first-1]){     //2n
            aux_product = hook;                                           //[0,n]
            execints++;
            m++;
        }
    }

    max_product = max_product * aux_product;                              //1
    execints++;

    std:: cout << "Answer: " << max_product << "\n";                      //1
    execints++;

    std::cout << "Cambios hechos: " << m << "\n";
                                                                            // -------------- 
    return max_product;                                                     // 6 + 6n + [0,n] 
}

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;                                 

    int n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    std::vector<long long> numbers(n);                             //n
    execints += n;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n
    }

    execints++;
    MaxPairwiseProduct(numbers,execints);      //1

    std::cout << "Execints: " << execints << "\n";        

    return 0;                                                // -------------- 
}                                                            // 10 + 10n + [0,n]

Overwriting max_pairwise_product_long_execints.cpp


Test with inputs 1 and 2



In [122]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!time ./max_pairwise_product_long_execints < input1.txt
!time ./max_pairwise_product_long_execints < input2.txt

Answer: 90
Cambios hechos: 10
Execints: 120

real	0m0.002s
user	0m0.002s
sys	0m0.000s
Answer: 40000000000
Cambios hechos: 5
Execints: 75

real	0m0.002s
user	0m0.001s
sys	0m0.001s


Test and time it with input 3

In [123]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!time ./max_pairwise_product_long_execints < input3.txt

Answer: 39999800000
Cambios hechos: 200000
Execints: 2200010

real	0m0.096s
user	0m0.085s
sys	0m0.009s


20 stress tests

In [109]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>


using namespace std;

long long MaxPairwiseProduct(const vector<long long>& numbers,long long execints) {
    long long max_product = 0LL;
    execints++;

	  long long aux_product = 0LL;
    execints++;

    int n = numbers.size();
    execints++;

    execints++;                                                                         
    for (int first = 0; first < n; ++first) {                           
        execints += 2;

        int hook = max_product;                                           
        execints++;

        max_product = std::max(max_product,numbers[first]);               
        execints++;

        execints++;
        execints++;
        if(hook != max_product || max_product == numbers[first-1]){     
            aux_product = hook;                                          
            execints++;
        }
    }

    return max_product;
}

int main() {
    long long execints =0LL;

    srand(time(NULL)); // inicializar semilla aleatoria
    execints++;

    execints++;
    for (int i = 0; i < 20; i++) {
        execints += 2;

        int n = rand() % 19 + 2; // tamaño aleatorio del arreglo entre [2, 20]
        execints++;

        vector<long long> numbers(n);
        execints += n;

        execints++;
        for (int j = 0; j < n; j++) {
            execints += 2;

            numbers[j] = rand() % 1000; // números aleatorios entre [0, 999]
            execints++;
        }
        MaxPairwiseProduct(numbers,execints);
        execints++;
    }

    cout << "Instrucciones: " << execints << endl;      
    return 0;
}


Overwriting test3.cpp


In [124]:
!g++ test3.cpp -o test3
!time ./test3

Instrucciones: 750

real	0m0.002s
user	0m0.000s
sys	0m0.002s


Count executed instructions in code and verify with inputs 1,2 and 3


In [130]:
print("Instrucciones input1: " + str(10+10*10+10))

Instrucciones input1: 120


In [127]:
print("Instrucciones input2: " + str(10+10*6+5))

Instrucciones input2: 75


In [128]:
print("Instrucciones input3: " + str(10+10*200000+200000))

Instrucciones input3: 2200010


Calculate estimated time by instruction and number of instructions by second

In [131]:
print(str(0.002/120) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son: "+ str(1/(0.002/120)))

1.6666666666666667e-05 Segundos por instruccion
La cantidad de instrucciones por segundo son: 60000.0


In [132]:
print(str(0.001/75) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son: "+ str(1/(0.001/75)))

1.3333333333333333e-05 Segundos por instruccion
La cantidad de instrucciones por segundo son: 75000.0


In [133]:
print(str(0.085/2200011) + "Segundos por instruccion")
print("La cantidad de instrucciones por segundo son: "+ str(1/(0.085/2200011)))

3.863617045551136e-08Segundos por instruccion
La cantidad de instrucciones por segundo son: 25882482.352941174


### Super efficient solution

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line 

In [146]:
%%writefile max_pairwise_product_long_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;                                

    long long n;                                                   //1
    execints++;                                                              
    
    std::cin >> n;                                           //1
    execints++;

    int number = 0LL;                                       //1
    execints++;

    int max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int m = 0LL;
    
    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> number;
        execints++;                                          //n

        int hook = max_product;                                           //n
        execints++;

        max_product = std::max(max_product,number);               //n
        execints++;

        execints++;
        execints++;
        if(hook != max_product || hook == number){               //2n
            aux_product = hook;                                           //[0,n]
            execints++;
            m++;
        }
    }
    
    std::cout << "Answer: " << max_product*aux_product << "\n";         //1
    execints++;

    std::cout << "Cambios hechos: " << m << "\n"; 

    std::cout << "Execints: " << execints << "\n";                            

    return 0;                                                // -------------- 
}                                                           //7 + 7n + [0,n]

Overwriting max_pairwise_product_long_execints.cpp


Test with inputs 1 and 2


 

In [147]:
!g++ max_pairwise_product_long_execints.cpp -o max_pairwise_product_long_execints
!time ./max_pairwise_product_long_execints < input1.txt
!time ./max_pairwise_product_long_execints < input2.txt

Answer: 90
Cambios hechos: 10
Execints: 87

real	0m0.003s
user	0m0.001s
sys	0m0.002s
Answer: 40000000000
Cambios hechos: 4
Execints: 53

real	0m0.002s
user	0m0.001s
sys	0m0.001s


Test and time it with input 3

In [148]:
!time ./max_pairwise_product_long_execints < input3.txt

Answer: 39999800000
Cambios hechos: 200000
Execints: 1600007

real	0m0.053s
user	0m0.049s
sys	0m0.003s


20 stress tests 

In [151]:
%%writefile test3.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>

int main() {                                                 //N inst exec by line executed 
    long long execints =0LL;                                
                                                    
    
    srand(time(NULL));                                                   //1
    execints++;          

    execints++;                                                           //1
    for(int i=0; i<20; i++){                                              //41
        execints += 2;

        int n = rand() % 19 + 2;                                          //1
        execints++;
        
        int max_product = 0LL;                                             //1
        execints++;

        int number = 0LL;                                                 //1
        execints++;

        long long aux_product = 0LL;                                       //1
        execints++;

        execints++;                                                        
        for (int j = 0; j < n; j++) {                                      //1+2n
            execints += 2;

            number = rand() % 1000;                                    //n
            execints++;

            int hook = max_product;                                           //n
            execints++;

            max_product = std::max(max_product,number);               //n
            execints++;

            execints++;
            execints++;
            if(hook != max_product || hook == number){               //2n
                aux_product = hook;                                           //[0,n]
                execints++;
            }
        }

    }                                                  
    std::cout << "Execints: " << execints << "\n";                            
    

    return 0;                                                // -------------- 
}          

Overwriting test3.cpp


In [153]:
!g++ test3.cpp -o test3
!time ./test3

Execints: 1668

real	0m0.002s
user	0m0.001s
sys	0m0.001s


Count executed instructions in code and verify with inputs 1,2 and 3

In [154]:
print("Instrucciones input1: " + str(7+(7*10)+10))

Instrucciones input1: 87


In [157]:
print("Instrucciones input2: " + str(7+(7*6)+4))

Instrucciones input2: 53


In [158]:
print("Instrucciones input2: " + str(7+(7*200000)+200000))

Instrucciones input2: 1600007


Calculate estimated time by instruction and number of instructions by second

In [159]:
print(str(0.003/87) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son en Input1: "+ str(1/(0.003/87)))

3.4482758620689657e-05 Segundos por instruccion
La cantidad de instrucciones por segundo son en Input1: 29000.0


In [160]:
print(str(0.002/53) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son en Input2: "+ str(1/(0.002/53)))

3.7735849056603776e-05 Segundos por instruccion
La cantidad de instrucciones por segundo son en Input2: 26500.0


In [161]:
print(str(0.053/1600007) + " Segundos por instruccion")
print("La cantidad de instrucciones por segundo son en Input3: "+ str(1/(0.053/1600007)))

3.3124855078759027e-08 Segundos por instruccion
La cantidad de instrucciones por segundo son en Input3: 30188811.32075472


# Codeforces problem https://codeforces.com/contest/1619/problem/A 

Solve the problem in Colab, submit the solution to Codeforces and add a screenshoot of the submission.

In [162]:
an = int(input())

for i in range(n):
    s = input()
    if len(s) % 2 == 0:
        half = len(s) // 2
        if s[:half] == s[half:]:
            print("YES")
        else: 
          print("NO")
    else:
      print("NO")

10
a
NO
aa
YES
aaa
NO
aaaa
YES
abab
YES
abcabc
YES
abacaba
NO
xxyy
NO
xyyx
NO
xyxy
YES
